CloudWatch infra support for eksapi deployer

annapoornanarayan opened the pull request:
*Issue #, if available:*

*Description of changes:*

Adds CloudWatch monitoring infrastructure setup to the eksapideployer, enabling automated deployment of CloudWatch agents and required IAM resources.

Files Modified:
- templates/templates.go - Added CloudWatch template definitions and data structures
- templates/cloudwatch-infra.yaml.template - OIDC provider and IAM role template
- templates/cloudwatch_agent_infra.yaml - Agent namespace, ClusterRole, and bindings
- cluster.go - Added OIDC issuer URL extraction
- infra.go - CloudWatch stack creation/deletion functions
- deployer.go - Integrated CloudWatch stacks after cluster and nodegroup formation and before deletion

Testing:
 - Verified CloudWatch infrastructure creates successfully
 - Confirmed proper cleanup on cluster deletion in Cloud Formation
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

wwvela commented:
Leave some nit comments, Overall LGTM

wwvela commented on internal/deployers/eksapi/infra.go:
```diff
@@ -475,3 +475,81 @@ func (m *InfrastructureManager) getAZsWithCapacity(opts *deployerOptions) ([]str
 }
+
+func (m *InfrastructureManager) createCloudWatchInfrastructureStack(clusterName string) (string, error) {
+	stackName := fmt.Sprintf("cloudwatch-%s", strings.TrimPrefix(clusterName, "kubetest2-eksapi-"))
+
+	var templateBuffer strings.Builder
+	err := templates.CloudWatchInfra.Execute(&templateBuffer, templates.CloudWatchInfraTemplateData{
+		ClusterName: clusterName,
+		ClusterUUID: strings.TrimPrefix(clusterName, "kubetest2-eksapi-"),
```
clusterUUID := strings.TrimPrefix(m.resourceID, "kubetest2-eksapi-")
as the uuid is used in multi place

wwvela commented on internal/deployers/eksapi/infra.go:
```diff
@@ -475,3 +475,81 @@ func (m *InfrastructureManager) getAZsWithCapacity(opts *deployerOptions) ([]str
+
+func (m *InfrastructureManager) deleteCloudWatchInfrastructureStack() error {
+	// We need to reconstruct the clusterUUID from resourceID to match the creation pattern
+	clusterUUID := strings.TrimPrefix(m.resourceID, "kubetest2-eksapi-")
+	stackName := fmt.Sprintf("cloudwatch-%s", clusterUUID)
```
uuid is only used in stack name, no need to assign to a param

mselim00 reviewed (commented):
thanks for putting this together!

annapoornanarayan commented on internal/deployers/eksapi/infra.go:
```diff
@@ -475,3 +475,80 @@ func (m *InfrastructureManager) getAZsWithCapacity(opts *deployerOptions) ([]str
 }
+
+func (m *InfrastructureManager) createCloudWatchInfrastructureStack(clusterName string) (string, error) {
+	clusterUUID := strings.TrimPrefix(clusterName, "kubetest2-eksapi-")
+	stackName := fmt.Sprintf("cloudwatch-%s", clusterUUID)
```
I changed it according to this comment: https://github.com/aws/aws-k8s-tester/pull/648#discussion_r2196121447

annapoornanarayan commented:
Thanks for the comments! Here is the summary of changes I made:
- Moved CW Role to Infrastructure struct instead of Cluster
- Created separate manifests folder
- Changed go substitution to CloudFormation string substitution
- Changed relevant warning logs to return errors in Cloudwatch infra creation and deletion
- Fixed recommended code style changes for code clarity

mselim00 commented on internal/deployers/eksapi/infra.go:
```diff
@@ -475,3 +476,74 @@ func (m *InfrastructureManager) getAZsWithCapacity(opts *deployerOptions) ([]str
 }
+
+func (m *InfrastructureManager) createCloudWatchInfrastructureStack(clusterName string) (string, error) {
+	clusterUUID := strings.TrimPrefix(clusterName, ResourcePrefix+"-")
+	stackName := fmt.Sprintf("cloudwatch-%s", clusterUUID)
```
nit: this is done twice, so it should be in a singular fuction

annapoornanarayan commented on internal/deployers/eksapi/templates/templates.go:
```diff
@@ -5,15 +5,21 @@ import (
 )
 
-//go:embed infra.yaml
+//go:embed manifests/infra.yaml
 var Infrastructure string
 
+//go:embed manifests/cloudwatch_agent_infra.yaml
+var CloudWatchAgentRbac []byte
+
```
So when I tried this, it was unable to access the file using` //go:embed` with relative paths as it does not support `..`. ([refer here)](https://pkg.go.dev/embed#:%7E:text=The%20patterns%20are%20interpreted%20relative%20to%20the%20package%20directory%20containing%20the%20source%20file.%20The%20path%20separator%20is%20a%20forward%20slash%2C%20even%20on%20Windows%20systems.%20Patterns%20may%20not%20contain%20%E2%80%98.%E2%80%99%20or%20%E2%80%98..%E2%80%99%20or%20empty%20path%20elements%2C%20nor%20may%20they%20begin%20or%20end%20with%20a%20slash.:~:text=The%20//go%3Aembed%20directive%20accepts%20multiple%20space%2Dseparated%20patterns%20for%20brevity%2C%20but%20it%20can%20also%20be%20repeated%2C%20to%20avoid%20very%20long%20lines%20when%20there%20are%20many%20patterns.%20The%20patterns%20are%20interpreted%20relative%20to%20the%20package%20directory%20containing%20the%20source%20file.) If i were to add a file outside the scope of templates.go I would have to create functions to embed.FS the entire directory. Its definitely possible, but it started to look like a large chunk of unnecessary code.

annapoornanarayan replied:
Actually you can ignore this comment, I think I found the issue!

annapoornanarayan replied:
Discussed offline and decided to move all files within templates
